File: bin/model_funcs/dfluxc.f
      subroutine dfluxc(ny,il,jl,
     & w,p,
     & porj,
     & fw, radi, radj,
c    no flo_params used
     & rfil,vis0)
c
c     
c     arg list above is:
c
c     dims
c
c     flow_var
c     mesh_var
c     solv_var
c
c     flo_param
c     solv_param
c
c
c     ******************************************************************
c     *                                                                *
c     *   artificial dissipation on coarse grids                       *
c     *   first order flux scaled to the spectral radius               *
c     *                                                                *
c     ******************************************************************
c
c     w(i,j,1)  = density
c     w(i,j,2)  = momentum in x direction
c     w(i,j,3)  = momentum in y direction
c     w(i,j,4)  = total energy
c
c     ******************************************************************
c
c      use dims
c
c     ******************************************************************
c
c      use flo_var
c      use mesh_var
c      use solv_var
c
c     ******************************************************************
c
c      use solv_param
c
c     ******************************************************************
c
      implicit none


c     ******************************************************************
c
c     input variables
c
c     ******************************************************************
c     from dims
      integer, intent(in) :: ny,il,jl

c     from flo_var
      real(8), intent(inout), dimension(:,:,:) :: w
      real(8), intent(inout), dimension(:,:) :: p

c     from mesh_var
      real(8), intent(in), dimension(:,:)   :: porj

c     from solv_var
      real(8), intent(inout), dimension(:,:,:) :: fw
      real(8), intent(inout), dimension(:,:)   :: radi,radj

c     from solv_param
      real(8), intent(in)     :: rfil, vis0
c
c     ******************************************************************
c
c     local variables
c
c     ******************************************************************
c
      integer  :: i,j,n
c
c     ******************************************************************
c
      real(8)     :: fis0,sfil
c
c     ******************************************************************
c
      real(8), dimension(il,jl,4)     :: fs
      real(8), dimension(il,jl)       :: dis
c
c     ******************************************************************
c
c      write(*,)
      fis0      = rfil*abs(vis0)/8.
      sfil      = 1.  -rfil
c
c     dissipation in the i direction
c
      do j=2,jl

         do i=1,il
c           dis(i,j)  = fis0*(radi(i+1,j)  +radi(i,j))
            dis(i,j)  = fis0*min(radi(i+1,j),radi(i,j))
         end do

         do n=1,4
            do i=1,il
               fs(i,j,n) = dis(i,j)*(w(i+1,j,n)  -w(i,j,n))
            end do
         end do

         do i=1,il
            fs(i,j,4) = fs(i,j,4)  +dis(i,j)*(p(i+1,j)  -p(i,j))
         end do

         do n=1,4
            do i=2,il
               fw(i,j,n) = sfil*fw(i,j,n)  -fs(i,j,n) +fs(i-1,j,n)
            end do
         end do

      end do

      if (ny.lt.3) return
c
c     dissipation in the j direction
c
      do j=1,jl
      do i=2,il
c        dis(i,j)  = fis0*(radj(i,j+1)  +radj(i,j))
         dis(i,j)  = fis0*porj(i,j)*min(radj(i,j+1),radj(i,j))
      end do
      end do

      do n=1,4
         do j=1,jl
         do i=2,il
            fs(i,j,n) = dis(i,j)*(w(i,j+1,n)  -w(i,j,n))
         end do
         end do
      end do

      do j=1,jl
      do i=2,il
         fs(i,j,4) = fs(i,j,4)  +dis(i,j)*(p(i,j+1)  -p(i,j))
      end do
      end do

      do n=1,4
         do j=2,jl
         do i=2,il
            fw(i,j,n) = fw(i,j,n)  -fs(i,j,n)  +fs(i,j-1,n)
         end do
         end do
      end do

      return

      end

File: bin/model_funcs/dfluxc.py
# python implementation of eflux.f

from bin.Field import Field, max, minimum
from bin.Grid import Grid
from bin.Workspace import Workspace
import numpy as np

def dfluxc(model, ws, state, dw, rfil):
    """
    calculate artificial dissipation fluxes on coarse meshes using blended first order 
    fluxes scaled to spectral radius
    
    Args:
        model (NavierStokes): physics model
        workspace (Workspace): contains the relevant Fields
        state (Field): density, x-momentum, y-momentum, and energy
        dw (Field): to store new residuals after completing fluxes 
        rfil (float): relaxation factor determining balance between viscous and artificial dissipation fluxes
        
    """
    # take a workspace ws and calculate dissipative fluxes

    # model parameters
    pad = model.padding
    n = Field.dim(state) # number of quantities being convected

    # necessary fields
    def mget(varName):
        return ws.get_field(varName, model.className)
    w = state # state vector
    fw   = mget('fw')
    p    = mget('p')   # pressure
    porJ = mget('porJ') # porosity
    radI = mget('radI')
    radJ = mget('radJ')

    # grid dimensions
    [nx, ny] = ws.field_size()
    [nxp, nyp] = [pad+nx+pad, pad+ny+pad]
    ip = pad
    jp = pad
    il = nx+1
    jl = ny+1
    ie = nx+2
    je = ny+2
    ib = nx+3
    jb = ny+3

    # physical paramteters
    vis0 = model.params['vis0']

    fis0      = rfil*np.abs(vis0).item()/8
    sfil      = 1.  -rfil

    # working array
    fs = Field.create((nxp, nyp, n))

    # c
    # c     dissipation in the i direction
    # c
    dis = minimum(radI[ip:ib, jp:je], radI[1:ie, jp:je])*fis0
    # dis = Field.create(dis)

    result = dis*(w[ip:ib, jp:je] - w[1:ie, jp:je])
    fs[1:ie, jp:je]     = result
    fs[1:ie, jp:je, 3] += dis*(p[ip:ib, jp:je] - p[1:ie, jp:je])

    fw[ip:ie, jp:je] = fw[ip:ie, jp:je]*sfil - fs[ip:ie, jp:je] + fs[1:il, jp:je]

    if ny < 3:
        return

    # c
    # c     dissipation in the j direction
    # c
    dis = porJ*fis0 * minimum(radI[ip:ie, jp:jb], radI[ip:ie, 1:je])

    fs[ip:ie, 1:je]     = dis*(w[ip:ie, jp:jb] - w[ip:ie, 1:je])
    fs[ip:ie, 1:je, 3] += dis*(p[ip:ie, jp:jb] - p[ip:ie, 1:je])

    fw[ip:ie, jp:je] += fs[ip:ie, jp:je] + fs[ip:ie, 1:jl]

    # add to dw
    dw += fw